pdns/dnsdist-web.cc

```diff
@@ -386,6 +386,120 @@ static void connectionThread(int sock, ComboAddress remote, string password, str
+          output << statesbase << "outstanding" << label << " " << state->outstanding.load() << "\n";
+        }
+
+        for (const auto& front : g_frontends) {
+          if (front->udpFD == -1 && front->tcpFD == -1)
+            continue;
+
+          string frontName = front->local.toString() + ":" + std::to_string(front->local.getPort());
+          boost::replace_all(frontName, ".", "_");
+          string proto = (front->udpFD >= 0 ? "udp" : "tcp");
+
+          output << "dnsdist_main_frontend_queries{frontend=\"" << frontName << "\",proto=\"" << proto
+              << "\"} " << front->queries.load() << "\n";
+        }
+
+        const auto localPools = g_pools.getCopy();
+        const string cachebase = "dnsdist_pool_";
+        
+        for (const auto& entry : localPools) {
+          string poolName = entry.first;
+          boost::replace_all(poolName, ".", "_");
+          
+          if (poolName.empty()) {
+            poolName = "_default_";
+          }
+          const string label = "{pool=\"" + poolName + "\"}";
+          const std::shared_ptr<ServerPool> pool = entry.second;
+          output << "dnsdist_main_pools_servers" << label << " " << pool->countServers(false) << "\n";
+
+          if (pool->packetCache != nullptr) {
+            const auto& cache = pool->packetCache;
+
+            output << cachebase << "cache_size"              <<label << " " << cache->getMaxEntries()       << "\n";
+            output << cachebase << "cache_entries"           <<label << " " << cache->getEntriesCount()     << "\n";
+            output << cachebase << "cache_hits"              <<label << " " << cache->getHits()             << "\n";
+            output << cachebase << "cache_misses"            <<label << " " << cache->getMisses()           << "\n";
+            output << cachebase << "cache_deferred_inserts"  <<label << " " << cache->getDeferredInserts()  << "\n";
+            output << cachebase << "cache_deferred_lookups"  <<label << " " << cache->getDeferredLookups()  << "\n";
+            output << cachebase << "cache_lookup_collisions" <<label << " " << cache->getLookupCollisions() << "\n";
+            output << cachebase << "cache_insert_collisions" <<label << " " << cache->getInsertCollisions() << "\n";
+            output << cachebase << "cache_ttl_too_shorts"    <<label << " " << cache->getTTLTooShorts()     << "\n";
+          }
+        }
+
+        {
+          WriteLock wl(&g_qcount.queryLock);
+          std::string qname;
+          const string qnamebase = "dnsdist_querycount_queries";
+
+          for(auto &record: g_qcount.records) {
+            qname = record.first;
+            boost::replace_all(qname, ".", "_");
+	    
+            const std::string label = "{qname=\"" + qname + "\"}";
+            output << qnamebase << label << " " << record.second << "\n";
+          }
+          g_qcount.records.clear();
```

Comment:
Are you sure you want to do that? Because right now the querycount stats are cleared after a carbon export, but with this change they would be cleared after a carbon export _and_ after an `API` query, which seems wrong. If you don't clear them then you only need a `ReadLock`, btw.

Reply:
Why is it reset after a carbon export? That will make these counters behave really weird, and graphs will show silly drops.

Reply:
I removed clear() here and replaced WriteLock by ReadLock. Also, I tested dnsdist with setQueryCount(true) and it works as expected.